### PR TITLE
Fix PromiseFulliller error when api::Socket is GC'd.

### PR DESCRIPTION
**Pull request by kentonv**
This turns out to be due to the `whenWriteDisconnected()` not being canceled before the connection stream is destroyed. Under KJ rules this would be a UAF, but in practice it harmlessly logs a spurious error: 'PromiseFulfiller was destroyed without fulfilling the promise.'

### CODE REVIEW

File: src/workerd/api/sockets.h [hunk 1]
```diff
@@ -89,13 +86,16 @@ class Socket: public jsg::Object {
...
 private:
+  // TODO(cleanup): Combine all the IoOwns here into one, to improve efficiency and make
+  //   shutdown order clearer.
+
   IoOwn<kj::RefcountedWrapper<kj::Own<kj::AsyncIoStream>>> connectionStream;
   jsg::Ref<ReadableStream> readable;
   jsg::Ref<WritableStream> writable;
-  jsg::PromiseResolverPair<void> closeFulfiller;
+  jsg::Promise<void>::Resolver closedResolver;
   // This fulfiller is used to resolve the `closedPromise` below.
```
Comment (dom96) [inline src/workerd/api/sockets.h:96]: Nit:
```suggestion
  // This resolver is used to resolve the `closedPromise` below.
```

File: src/workerd/api/sockets.c++ [hunk 2]
```diff
@@ -70,7 +70,63 @@ jsg::Ref<Socket> setupSocket(
...
+  // `watchForDisconnectTask`, which is a `kj::Promise<void>` representing a task that waits for
+  // `whenWriteDisconnected()` and then fulfills the fulfiller end of `disconnectedPaf` with
+  // `false`. If the task is canceled, we instead fulfill `disconnectedPaf` with `true`.
+  //
+  // We then use `context.awaitIo()` to await the promise end of `disconnectedPaf`, and this gives
+  // us our `closed` promise. Well, almost...
+  //
+  // There's another wrinkle: There are some circumstances where we want to resolve the `closed`
+  // promise directly from an API call. We'd rather this did not have to drop out of the isolate
+  // and enter it a gain. So, our `awaitIo()` actually awaits a task that listens for the
```
Comment (dom96) [inline src/workerd/api/sockets.c++:94]: Nit:
```suggestion
  // and enter it again. So, our `awaitIo()` actually awaits a task that listens for the
```

**Review by dom96 — APPROVED** [review]

**Comment by alanxone** [comment]
When will the fix be deployed?

**Comment by kentonv** [comment]
@alanxone Err, why do you ask? This is fixing some spurious errors in our internal monitoring, it shouldn't be a user-visible change.

**Pull request merged**